### PR TITLE
Threat objects

### PR DESCRIPTION
**Pull request by ljstella**
Supersedes #204 

Breaks out risk objects vs threat objects for the purpose of "what needs to be in every result" vs "what needs to be have a value at least once". Instead of trying to keep track of what's missing, its much easier to keep track of what is definitely **not** null, adding it to a set as we go, and comparing it to the full set at the end. 

@pyth0n1c feel free to test with all sorts of broken YAMLs, I *think* I have accounted for all of the weird edge cases of potential test failures, but its also possible I missed something between banging my head on the keyboard and wishing for magical datatypes that don't exist.

@cmcginley-splunk likewise, feel free to test. I _tried_ to keep up with the type annotations and formatting elsewhere in this, but its quite likely I've ruined some of the stuff you've added.

### CODE REVIEW

**Comment by ljstella** [comment]
Forgot to note, you can use https://github.com/splunk/security_content/pull/2995 as an example PR that currently fails against `main` but passes with this- the detection `detections/cloud/o365_zap_activity_detection.yml` in particular exhibits the specific behavior I set out to check against- threat objects that are not present in all risk events, but are present in at least one. I tested we can still fail when something isn't present at all by adding a `values(filename) as filename` into the search and adjusting the observable `file_name` to `filename`- that field isn't in the underlying data so always ends up null.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py [hunk 1]
```diff
@@ -299,6 +299,11 @@ def risk(self) -> list[dict[str, Any]]:
                 risk_object['threat_object_field'] = entity.name
                 risk_object['threat_object_type'] = "url"
                 risk_objects.append(risk_object)
+            
+            elif 'Attacker' in entity.role:
+                risk_object['threat_object_field'] = entity.name
+                risk_object['threat_object_field'] = entity.type.lower()
```
Comment (pyth0n1c) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:305]: Should this actually be `risk_object['threat_object_type'] = entity.type.lower()` ?

**Review by pyth0n1c — CHANGES_REQUESTED** [review]
I have left some comments with questions below, let me know your thoughts @ljstella .

I also wanted to post a few points for clarity:
1. It would be FAR better if we had improved Pydantic Objects/typing around the "observables" section of our YMLs.  Having this be a mix between Threat Objects and Risk Objects, which have different behaviors, leads to some confusion and far more complex handing in functions like this. I understand that updating those objects is a larger change and outside the scope of this PR.

2.  These changes update the check on observables to require that:
a.  If an observable is a Risk Object, EVERY result must contain a non-null value for each Risk Object.
b. If an observable is a THREAT object, AT LEAST 1 of the results must contain a non-null value for each Threat Object. Here an is example: https://github.com/splunk/security_content/pull/2995/files#diff-93a083d42a51945f1ae116446dba1974ba182e98a18587ab3976b6054e797916R40-R51
and a screenshot using that detection/test data:
image

Based on our discussions, we made this change because multiple different threats can create risk for one of my assets.  In order to account for these different Threats, we allow one threat per event (but enforce that all possible Threat Objects must be created during a test/test dataset)

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py [hunk 2]
```diff
@@ -1060,29 +1060,36 @@ def retry_search_until_timeout(
...
+            threat_object_fields_set = set([o.name for o in detection.tags.observable if "Attacker" in o.role]) # just the "threat objects"
 
             # Ensure the search had at least one result
             if int(job.content.get("resultCount", "0")) > 0:
                 # Initialize the test result
                 test.result = UnitTestResult()
 
                 # Initialize the collection of fields that are empty that shouldn't be
+                present_threat_objects: set[str] = set()
                 empty_fields: set[str] = set()
```
[the remark this answers is not in the document]
Comment (ljstella) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py:1074]: on line 1144 , this set is used to determine the missing ones with some set arithmetic

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py [hunk 3]
```diff
@@ -1114,21 +1140,32 @@ def retry_search_until_timeout(
 
                     else:
                         empty_fields = empty_fields.union(current_empty_fields)
```
[the remark this answers is not in the document]
Comment (ljstella) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py:1142]: I was keeping the `empty_fields` set around specifically because it was handy to check with the debugger while writing this & I was hesitant to remove what appeared to be load bearing. We can take it out if you want, but its a minimal amount of overhead computationally.

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py [hunk 4]
```diff
@@ -1060,29 +1060,36 @@ def retry_search_until_timeout(
...
-                    # Identify any observable fields that are not available in the results
-                    missing_fields = observable_fields_set - results_fields_set
-                    if len(missing_fields) > 0:
+                    # Identify any risk object fields that are not available in the results
+                    missing_risk_objects = risk_object_fields_set - results_fields_set
+                    if len(missing_risk_objects) > 0:
                         # Report a failure in such cases
-                        e = Exception(f"The observable field(s) {missing_fields} are missing in the detection results")
+                        e = Exception(f"The observable field(s) {missing_risk_objects} are missing in the detection results")
                         test.result.set_job_content(
```
[the remark this answers is not in the document]
Comment (ljstella) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py:1092]: Cleaned this up

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py (hunk 1 above)
Comment (ljstella) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py]: Yup, thanks!

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pyth0n1c) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py:1074]: This was supposed to refer to empty_fields. I saw your comment below about keeping it around for debugging, that's okay :) We will keep it.

**Review by pyth0n1c — APPROVED** [review]
Thanks Lou for feedback on my comments. After additional local testing, this looks great! 
Approved!

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py [hunk 5]
```diff
@@ -299,6 +299,11 @@ def risk(self) -> list[dict[str, Any]]:
                 risk_object['threat_object_field'] = entity.name
                 risk_object['threat_object_type'] = "url"
                 risk_objects.append(risk_object)
+            
+            elif 'Attacker' in entity.role:
```
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:303]: Note that I tested an ESCU build made using this branch against the new testing in my branch (see #241). My risk/observable matching showed 50+ detections generating risk events for Attacker observables, which shouldn't happen. These changes in your PR Lou resolved all 50+ of those detections.
